tox.ini

```diff
@@ -18,73 +18,50 @@ passenv =
   PYTHONPATH
   CHARM_BUILD_DIR
   MODEL_SETTINGS
+allowlist_externals = uv
+
+[testenv:generate-requirements]
+description = Generate requirements files
+commands =
+    uv pip compile -q --no-cache pyproject.toml -o requirements.txt
+    uv pip compile -q --no-cache --all-extras pyproject.toml -o requirements-dev.txt
 
 [testenv:fmt]
 description = Apply coding style standards to code
 deps =
-    # renovate: datasource=pypi
-    ruff==0.6.2
+    -r {toxinidir}/requirements-dev.txt
 commands =
-    ruff check --fix {[vars]all_path}
-    ruff format {[vars]all_path}
+    uv run ruff check --fix {[vars]all_path}
+    uv run ruff format {[vars]all_path}
```

Comment:
My 2c: these should be declared as tools, as those get their own venv:

https://docs.astral.sh/uv/guides/tools/

Full disclosure: I've only scratched the tool API surface, and I'm not sure how they work, really.

It seems that tools are installed into local and are not versioned in pyproject.toml?

Reply:
Done, now using `uvx`, which reading the docs seems more appropriate.